garden-backend-service/src/api/routes/gardens.py

```diff
@@ -109,3 +62,108 @@ async def delete_garden(
         return {"detail": f"Successfully deleted garden with DOI {doi}."}
     else:
         return {"detail": f"No garden found with DOI {doi}."}
+
+
+@router.put("/{doi:path}", response_model=GardenMetadataResponse)
+async def create_or_replace_garden(
+    doi: str,
+    garden_data: GardenCreateRequest,
+    db: AsyncSession = Depends(get_db_session),
+    user: User = Depends(authed_user),
+):
+    existing_garden: Garden | None = await Garden.get(db, doi=doi)
+    if existing_garden is None:
+        return await _create_new_garden(garden_data, db, user)
+
+    # check draft status with the real world (doi.org)
+    if garden_data.doi_is_draft is None:
+        registered = await is_doi_registered(garden_data.doi)
+        garden_data.doi_is_draft = not registered
+
+    assert_deletable_by_user(existing_garden, user)
+    # update ownership if specified
+    if garden_data.owner_identity_id is not None:
+        new_owner: User | None = await User.get(
+            db, identity_id=garden_data.owner_identity_id
+        )
+        existing_garden.owner = new_owner or user
+
+    # update related entrypoints
+    new_entrypoints = await _collect_entrypoints(garden_data.entrypoint_ids, db)
+    existing_garden.entrypoints = new_entrypoints
+
+    # naive update with remaining values from payload
+    for key, value in garden_data.model_dump(
+        exclude={"owner_identity_id", "entrypoint_ids"}
+    ).items():
+        setattr(existing_garden, key, value)
+    try:
+        await db.commit()
+    except IntegrityError as e:
+        logger.error(str(e))
+        await db.rollback()
+        raise HTTPException(
+            status_code=status.HTTP_409_CONFLICT,
+            detail=f"Integrity error occurred: {str(e)}",
+        ) from e
+
+    return existing_garden
+
+
+async def _collect_entrypoints(dois: list[str], db: AsyncSession) -> list[Entrypoint]:
+    stmt = select(Entrypoint).where(Entrypoint.doi.in_(dois))
+    result = await db.execute(stmt)
+    entrypoints: list[Entrypoint] = result.scalars().all()
+
+    if len(entrypoints) != len(dois):
+        missing_dois = [ep.doi for ep in entrypoints if ep.doi not in dois]
+        raise HTTPException(
+            status_code=404,
+            detail=f"Could not find entrypoint(s) with DOIs: {missing_dois}",
+        )
+    return entrypoints
+
+
+async def _create_new_garden(
+    garden_data: GardenCreateRequest,
+    db: AsyncSession,
+    user: User,
+):
+    # if not specified, check draft status with the real world (doi.org)
+    if garden_data.doi_is_draft is None:
+        registered = await is_doi_registered(garden_data.doi)
+        garden_data.doi_is_draft = not registered
+
+    # collect entrypoints by DOI
+    entrypoints = await _collect_entrypoints(garden_data.entrypoint_ids, db)
```

Comment:
I'm curious, why do we need the fully hydrated Entrypoint objects to create a Garden? It makes sense that we need to check if the entrypoint ids the user provided correspond to entrypoints in the DB. But why do we need to stick the full objects on to `new_garden`?

Reply:
Oh I might have a guess ... in the 201 CREATED response you want to send back the whole garden including its full entrypoint metadata in the response. So that's why it's useful to grab the full Entrypoint objects. (maybe?)

Reply:
assigning the full entrypoint objects to the garden was honestly just the first thing I attempted that got sqlalchemy to update the many-many association table correctly. definitely interested if I was missing a better way to do that with the ORM